File: forms/hospital-discharge/back-end-with-loco/src/engine/discharge_validator.rs
use super::flagged_issues::detect_additional_flags;
use super::types::{AssessmentData, CompletenessLevel, FiredRule, GradingResult};
use super::validation_rules::all_rules;

/// Pure function: grades a hospital discharge summary against the NICE
/// NG27 mandatory and optional rule set.
///
/// Completeness cutoffs:
/// - All mandatory rules satisfied + all optional rules satisfied -> `complete`
/// - All mandatory rules satisfied + some optional rules missing  -> `partial`
/// - Any mandatory rule unsatisfied                                 -> `incomplete`
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let (
        completeness_level,
        mandatory_satisfied,
        mandatory_total,
        optional_satisfied,
        optional_total,
        fired_rules,
    ) = validate_discharge(data);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        completeness_level,
        mandatory_satisfied,
        mandatory_total,
        optional_satisfied,
        optional_total,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

/// Evaluate every NG27 rule and tally satisfied / total counts.
pub fn validate_discharge(
    data: &AssessmentData,
) -> (CompletenessLevel, u32, u32, u32, u32, Vec<FiredRule>) {
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut mandatory_satisfied: u32 = 0;
    let mut mandatory_total: u32 = 0;
    let mut optional_satisfied: u32 = 0;
    let mut optional_total: u32 = 0;

    for rule in all_rules() {
        let satisfied = (rule.evaluate)(data);
        if rule.mandatory {
            mandatory_total += 1;
            if satisfied {
                mandatory_satisfied += 1;
            }
        } else {
            optional_total += 1;
            if satisfied {
                optional_satisfied += 1;
            }
        }
        fired_rules.push(FiredRule {
            id: rule.id.to_string(),
            category: rule.category.to_string(),
            description: rule.description.to_string(),
            satisfied,
            mandatory: rule.mandatory,
        });
    }

    let completeness_level: CompletenessLevel = if mandatory_satisfied < mandatory_total {
        "incomplete".to_string()
    } else if optional_satisfied < optional_total {
        "partial".to_string()
    } else {
        "complete".to_string()
    };

    (
        completeness_level,
        mandatory_satisfied,
        mandatory_total,
        optional_satisfied,
        optional_total,
        fired_rules,
    )
}

File: forms/hospital-discharge/front-end-with-svelte/src/lib/engine/discharge-validator.ts
// Discharge summary completeness validator. Pure functions: take an
// `AssessmentData` object, return the per-rule audit trail and the
// overall completeness level.
//
// Completeness cutoffs (NICE NG27):
//   All mandatory rules satisfied + all optional rules satisfied -> Complete
//   All mandatory rules satisfied + some optional rules missing  -> Partial
//   Any mandatory rule unsatisfied                                -> Incomplete

import type { AssessmentData, CompletenessLevel, FiredRule, GradingResult } from './types';
import { validationRules } from './validation-rules';
import { detectAdditionalFlags } from './flagged-issues';

/** The aggregated outcome of running every NICE NG27 validation rule. */
export interface ValidationSummary {
	completenessLevel: CompletenessLevel;
	mandatorySatisfied: number;
	mandatoryTotal: number;
	optionalSatisfied: number;
	optionalTotal: number;
	firedRules: FiredRule[];
}

/** Validate the discharge summary against all NICE NG27 rules. */
export function validateDischarge(data: AssessmentData): ValidationSummary {
	const firedRules: FiredRule[] = [];
	let mandatorySatisfied = 0;
	let mandatoryTotal = 0;
	let optionalSatisfied = 0;
	let optionalTotal = 0;

	for (const rule of validationRules) {
		let satisfied = false;
		try {
			satisfied = rule.evaluate(data) === true;
		} catch (e) {
			console.warn(`Validation rule ${rule.id} evaluation failed:`, e);
			satisfied = false;
		}
		if (rule.mandatory) {
			mandatoryTotal++;
			if (satisfied) mandatorySatisfied++;
		} else {
			optionalTotal++;
			if (satisfied) optionalSatisfied++;
		}
		firedRules.push({
			id: rule.id,
			category: rule.category,
			description: rule.description,
			mandatory: rule.mandatory,
			satisfied
		});
	}

	let completenessLevel: CompletenessLevel;
	if (mandatorySatisfied < mandatoryTotal) {
		completenessLevel = 'incomplete';
	} else if (optionalSatisfied < optionalTotal) {
		completenessLevel = 'partial';
	} else {
		completenessLevel = 'complete';
	}

	return {
		completenessLevel,
		mandatorySatisfied,
		mandatoryTotal,
		optionalSatisfied,
		optionalTotal,
		firedRules
	};
}

/**
 * Pure grader entry point: validate the discharge summary and attach the
 * clinician-facing flags, producing the full report payload.
 */
export function gradeDischarge(data: AssessmentData): GradingResult {
	const validation = validateDischarge(data);
	return {
		...validation,
		additionalFlags: detectAdditionalFlags(data),
		timestamp: new Date().toISOString()
	};
}
